### PR TITLE
feat: support logical evaluation

### LINKED ISSUE #5038: Compile error caused by a node module
### System Info

```
  System:
    OS: macOS 14.2
    CPU: (10) arm64 Apple M1 Pro
    Memory: 94.67 MB / 16.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 20.10.0 - ~/Library/Caches/fnm_multishells/76248_1702736941772/bin/node
    Yarn: 3.3.1 - ~/Library/Caches/fnm_multishells/76248_1702736941772/bin/yarn
    npm: 10.2.3 - ~/Library/Caches/fnm_multishells/76248_1702736941772/bin/npm
    pnpm: 8.12.1 - ~/Library/Caches/fnm_multishells/76248_1702736941772/bin/pnpm
    bun: 1.0.6 - ~/.bun/bin/bun
  Browsers:
    Chrome: 120.0.6099.109
    Safari: 17.2
```

### Details

Key words:

- resolver
- node module

Related codes:

```JavaScript
export async function foo(input) {
  switch (true) {
    case typeof input === 'string':
      input = new URL(input, import.meta.url);
    case input instanceof URL:
      if (typeof __webpack_require__ !== 'function' && input.protocol === 'file:') {
        const fs = await import('node:fs/promises');
        input = fs.readFile(input);
        break;
      }
    case typeof Request === 'function' && input instanceof Request:
      input = fetch(input);
  }
}
```

Real-world issue:

This code snippet is using in @wasm-fmt/ruff_fmt .
Therefore, importing this library may result in compilation errors.

### Reproduce link

_No response_

### Reproduce Steps

1.  Create project via cli using Vanilla template
```
❯ pnpm create rsbuild@latest

◆  Create Rsbuild Project
│
◇  Input target folder
│  my-project
│
◇  Select framework
│  Vanilla
│
◇  Select language
│  JavaScript
│
◇  Next steps ────╮
│                 │
│  cd my-project  │
│  pnpm i         │
│  pnpm run dev   │
│                 │
├─────────────────╯
│
└  Done.
```

2. Replace the content in `src/index.js` with the code snippet mentioned above.
3. pnpm i
4. pnpm build

You will get the error
```

> rsbuild-vanilla-js@1.0.0 build /Users/akari/Developer/my-project
> rsbuild build

  Rsbuild v0.2.5

start   Use Rspack v0.4.3
● Client ━━━━━━━━━━━━━━━━━━━━━━━━━ (100%) done                                                                                                                             error   Compile error: 
Failed to compile, check the errors for troubleshooting.
× No such file or directory (os error 2), failed to read

error   Failed to build.
error   Rspack build failed!
    at /Users/akari/Developer/my-project/node_modules/.pnpm/@rsbuild+core@0.2.5/node_modules/@rsbuild/core/dist/provider/core/build.js:32:35
    at finalCallback (/Users/akari/Developer/my-project/node_modules/.pnpm/@rspack+core@0.4.3/node_modules/@rspack/core/dist/Compiler.js:335:21)
    at /Users/akari/Developer/my-project/node_modules/.pnpm/@rspack+core@0.4.3/node_modules/@rspack/core/dist/Compiler.js:359:40
    at Hook.eval [as callAsync] (eval at create (/Users/akari/Developer/my-project/node_modules/.pnpm/tapable@2.2.1/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:24:1)
    at Hook.CALL_ASYNC_DELEGATE [as _callAsync] (/Users/akari/Developer/my-project/node_modules/.pnpm/tapable@2.2.1/node_modules/tapable/lib/Hook.js:18:14)
    at /Users/akari/Developer/my-project/node_modules/.pnpm/@rspack+core@0.4.3/node_modules/@rspack/core/dist/Compiler.js:354:41
    at /Users/akari/Developer/my-project/node_modules/.pnpm/@rspack+core@0.4.3/node_modules/@rspack/core/dist/Compiler.js:391:21
 ELIFECYCLE  Command failed with exit code 1.
```

### PR DESCRIPTION
**Pull request by LingyuCoder**
## Summary

Fix #5038

## Test Plan

- Add `packages/rspack/tests/cases/parsing/evaluate-logical`

## Require Documentation?

- [x] No
- [ ] Yes, the corresponding rspack-website PR is \_\_

### CODE REVIEW

**Comment by LingyuCoder** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/web-infra-dev/rspack/5162?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `main`
  * **PR #5161** <a href="https://app.graphite.dev/github/pr/web-infra-dev/rspack/5161?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #5162** <a href="https://app.graphite.dev/github/pr/web-infra-dev/rspack/5162?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/web-infra-dev/rspack/5162?utm_source=stack-comment).

File: crates/rspack_plugin_javascript/src/utils/eval/eval_binary_expr.rs [hunk 1]
```diff
@@ -117,6 +117,87 @@ fn handle_abstract_equality_comparison(
   }
 }
 
+fn handle_logical_or(
```
Comment (bvanjoi) [inline crates/rspack_plugin_javascript/src/utils/eval/eval_binary_expr.rs:120]: I believe this PR duplicates the content in `﻿rspack_plugin_javascript/src/utils/const/logic_expr`

Comment (LingyuCoder) [reply crates/rspack_plugin_javascript/src/utils/eval/eval_binary_expr.rs:120]: Actually they have some differences, such as the following example:

```js
// source code, false && true = false
if (typeof require !== 'function' && typeof require === 'function') { 
  require('fail');
}

// webpack, will remove dead code
if (false) {}

// rspack before, will calcualte each side but not a whole result, then dead code will be parsed
if ( false && 0) __webpack_require__(Object(function webpackMissingModule() { var e = new Error("Cannot find module 'fail'"); e.code = 'MODULE_NOT_FOUND'; throw e; }()));

// rspack after, will remove just like webpack
if (false) {}
```

Comment (bvanjoi) [reply crates/rspack_plugin_javascript/src/utils/eval/eval_binary_expr.rs:120]: Oh, I see. There won't be any insertion of `﻿ConstDependency`

**Review by bvanjoi — DISMISSED** [review]

**Review by bvanjoi — APPROVED** [review]

**Pull request merged**